File: src/main/java/ThirtyDayLeetCodingChallenge/Day25_JumpGame.java
package ThirtyDayLeetCodingChallenge;

public class Day25_JumpGame {
    public static void main(String[] args) {
        System.out.println(canJump(new int[]{2, 3, 1, 1, 4}));
        System.out.println(canJump(new int[]{3, 2, 1, 0, 4}));
        System.out.println(canJump(new int[]{4, 3, 2, 2, 0, 0, 4}));
    }

    public static boolean canJump(int[] nums) {
        int currentIndex = 0;
        for (int i = 0; i < nums.length && i <= currentIndex; i++) {
            currentIndex = Math.max(i + nums[i], currentIndex);
        }

        return currentIndex >= nums.length - 1;
    }
}

File: src/main/kotlin/ThirtyDayLeetCodingChallenge/Day25_JumpGame.kt
import kotlin.math.max

fun main() {
    println(canJump(intArrayOf(2, 3, 1, 1, 4)))
    println(canJump(intArrayOf(3, 2, 1, 0, 4)))
}

fun canJump(nums: IntArray): Boolean {
    var currentIndex = 0
    for (i in nums.indices) {
        if (i > currentIndex) return false
        currentIndex = max(i + nums[i], currentIndex)
    }

    return true
}
